docs/smart-contracts/meebits_erc721.sol

```diff
@@ -0,0 +1,679 @@
+        }
+        nonce++;
+        // Don't allow a zero index, start counting at 1
+        return value.add(1);
+    }
+
+    // Calculate the mint price
+    function getPrice() public view returns (uint) {
+        require(publicSale, "Sale not started.");
+        uint elapsed = block.timestamp.sub(saleStartTime);
+        if (elapsed >= saleDuration) {
+            return 0;
+        } else {
+            return saleDuration.sub(elapsed).mul(price).div(saleDuration);
+        }
+    }
+
+    // The deployer can mint in bulk without paying
+    function devMint(uint quantity, address recipient) external onlyDeployer {
+        for (uint i = 0; i < quantity; i++) {
+            _mint(recipient, 0);
+        }
+    }
+
+    function mintsRemaining() external view returns (uint) {
+        return SALE_LIMIT.sub(numSales);
+    }
+
+    /**
+     * Community grant minting.
+     */
+    function mintWithPunkOrGlyph(uint _createVia) external reentrancyGuard returns (uint) {
+        require(communityGrant);
+        require(!marketPaused);
+        require(_createVia > 0 && _createVia <= 10512, "Invalid punk/glyph index.");
+        require(creatorNftMints[_createVia] == 0, "Already minted with this punk/glyph");
+        if (_createVia > 10000) {
+            // It's a glyph
+            // Compute the glyph ID
+            uint glyphId = _createVia.sub(10000);
+            // Make sure the sender owns the glyph
+            require(IERC721(glyphs).ownerOf(glyphId) == msg.sender, "Not the owner of this glyph.");
+        } else {
+            // It's a punk
+            // Compute the punk ID
+            uint punkId = _createVia.sub(1);
+            // Make sure the sender owns the punk
+            require(Cryptopunks(punks).punkIndexToAddress(punkId) == msg.sender, "Not the owner of this punk.");
+        }
+        creatorNftMints[_createVia]++;
+        return _mint(msg.sender, _createVia);
+    }
+
+    /**
+     * Public sale minting.
+     */
+    function mint() external payable reentrancyGuard returns (uint) {
+        require(publicSale, "Sale not started.");
+        require(!marketPaused);
+        require(numSales < SALE_LIMIT, "Sale limit reached.");
+        uint salePrice = getPrice();
+        require(msg.value >= salePrice, "Insufficient funds to purchase.");
+        if (msg.value > salePrice) {
+            msg.sender.transfer(msg.value.sub(salePrice));
+        }
+        beneficiary.transfer(salePrice);
+        numSales++;
+        return _mint(msg.sender, 0);
+    }
+
+    function _mint(address _to, uint createdVia) internal returns (uint) {
+        require(_to != address(0), "Cannot mint to 0x0.");
+        require(numTokens < TOKEN_LIMIT, "Token limit reached.");
+        uint id = randomIndex();
+
+        numTokens = numTokens + 1;
+        _addNFToken(_to, id);
+
+        emit Mint(id, _to, createdVia);
+        emit Transfer(address(0), _to, id);
+        return id;
+    }
```

Comment:
Sends a token based on a random index 
Emits a Mint event
Emits a ERC721.Transfer event from null address and to recipient